trieste/models/model_interfaces.py

```diff
@@ -94,6 +94,95 @@ def optimize(self, dataset: Dataset) -> None:
+            shape [..., :math:`\sum_i E_i`].
+        """
+        means, vars_ = zip(*[model.predict(query_points) for model in self._models])
+        return tf.concat(means, axis=-1), tf.concat(vars_, axis=-1)
+
+    def predict_joint(self, query_points: TensorType) -> tuple[TensorType, TensorType]:
+        r"""
+        :param query_points: The points at which to make predictions, of shape [..., B, D].
+        :return: The predictions from all the wrapped models, concatenated along the event axis in
+            the same order as they appear in :meth:`__init__`. If the wrapped models have predictive
+            distributions with event shapes [:math:`E_i`], the mean will have shape
+            [..., B, :math:`\sum_i E_i`], and the covariance shape
+            [..., :math:`\sum_i E_i`, B, B].
+        """
+        means, covs = zip(*[model.predict_joint(query_points) for model in self._models])
+        return tf.concat(means, axis=-1), tf.concat(covs, axis=-3)
+
+    def sample(self, query_points: TensorType, num_samples: int) -> TensorType:
+        r"""
+        :param query_points: The points at which to sample, with shape [..., D].
+        :param num_samples: The number of samples at each point.
+        :return: The samples from all the wrapped models, concatenated along the event axis. For
+            wrapped models with predictive distributions with event shapes [:math:`E_i`], this has
+            shape [..., S, :math:`\sum_i E_i`], where S is the number of samples.
+        """
+        samples = [model.sample(query_points, num_samples) for model in self._models]
+        return tf.concat(samples, axis=-1)
+
+    def update(self, dataset: Dataset) -> None:
+        """
+        Update all the wrapped models on their corresponding data. The data for each model is
+        extracted by splitting the observations in ``dataset`` along the event axis according to the
+        event sizes specified at :meth:`__init__`.
+
+        :param dataset: The query points and observations for *all* the wrapped models.
+        """
+        observations = tf.split(dataset.observations, self._event_sizes, axis=-1)
+
+        for model, obs in zip(self._models, observations):
+            model.update(Dataset(dataset.query_points, obs))
+
+    def optimize(self, dataset: Dataset) -> None:
+        """
+        Optimize all the wrapped models on their corresponding data. The data for each model is
+        extracted by splitting the observations in ``dataset`` along the event axis according to the
+        event sizes specified at :meth:`__init__`.
+
+        :param dataset: The query points and observations for *all* the wrapped models.
+        """
+        observations = tf.split(dataset.observations, self._event_sizes, axis=-1)
+
+        for model, obs in zip(self._models, observations):
+            model.optimize(Dataset(dataset.query_points, obs))
```

Comment:
Not needed for this PR, but I think we should allow a bit more flexibility over the model optimiser here. i.e similar to how we discussed with the acquisition optimiser.  It is very often the case that we might want to at least control the number of random restarts, especially in the early BO steps where there is little data.

Reply:
the optimizer is configurable for each individual model. See the `optimizer` argument to the constructors for the GPflow model wrappers. Is this what you mean?